File: backend/twitter-service/src/rabbitmq/publisher.rs
use lapin::{
    options::{BasicPublishOptions, ExchangeDeclareOptions},
    types::FieldTable,
    BasicProperties, Channel, Connection, ConnectionProperties,
};

/// Connect to RabbitMQ, create a channel, and declare the `ecoguard.events` topic exchange.
pub async fn connect(uri: &str) -> Result<Channel, Box<dyn std::error::Error>> {
    let conn = Connection::connect(uri, ConnectionProperties::default()).await?;
    let channel = conn.create_channel().await?;

    channel
        .exchange_declare(
            "ecoguard.events",
            lapin::ExchangeKind::Topic,
            ExchangeDeclareOptions {
                durable: true,
                ..Default::default()
            },
            FieldTable::default(),
        )
        .await?;

    println!("✅ Connected to RabbitMQ, exchange ecoguard.events ready");
    Ok(channel)
}

/// Publish a JSON payload to the `tweet.ingested` routing key.
pub async fn publish_tweet_ingested(
    channel: &Channel,
    payload: &[u8],
) -> Result<(), Box<dyn std::error::Error>> {
    channel
        .basic_publish(
            "ecoguard.events",
            "tweet.ingested",
            BasicPublishOptions::default(),
            payload,
            BasicProperties::default(),
        )
        .await?;
    Ok(())
}

File: backend/issue-service/rabbitmq/publisher.py
"""
RabbitMQ event publisher untuk Issue Service.
Publish event ke exchange ecoguard.events (topic).
"""

import json
import logging

import pika

logger = logging.getLogger(__name__)


class EventPublisher:
    """Publish issue events (resolve, create) to RabbitMQ event bus."""

    def __init__(self, rabbitmq_uri: str):
        self.uri = rabbitmq_uri
        self.connection = None
        self.channel = None
        self._connect()

    def _connect(self):
        try:
            params = pika.URLParameters(self.uri)
            params.heartbeat = 300  # 5 menit heartbeat
            params.blocked_connection_timeout = 30
            self.connection = pika.BlockingConnection(params)
            self.channel = self.connection.channel()
            self.channel.exchange_declare(
                exchange="ecoguard.events",
                exchange_type="topic",
                durable=True,
            )
            logger.info("Connected to RabbitMQ")
        except Exception as e:
            logger.warning("RabbitMQ connection failed: %s", e)
            self.connection = None
            self.channel = None

    def _ensure_connected(self):
        """Reconnect if connection is closed/stale."""
        if self.connection and self.connection.is_open:
            return True
        logger.info("RabbitMQ connection lost, reconnecting...")
        self._connect()
        return self.channel is not None

    def publish_issue_resolved(
        self, issue_id: str, tweet_id: str, admin_id: str, notes: str,
        image_hashes: list[str], resolved_at: int,
    ):
        """Publish issue.resolved event."""
        if not self._ensure_connected():
            return
        body = json.dumps({
            "event": "issue.resolved",
            "issue_id": issue_id,
            "tweet_id": tweet_id,
            "admin_id": admin_id,
            "notes": notes,
            "image_hashes": image_hashes,
            "resolved_at": resolved_at,
        })
        try:
            self.channel.basic_publish(
                exchange="ecoguard.events",
                routing_key="issue.resolved",
                body=body,
                properties=pika.BasicProperties(delivery_mode=2),
            )
            logger.info("Published issue.resolved: %s", issue_id)
        except Exception as e:
            logger.warning("Failed to publish event: %s", e)

    def close(self):
        if self.connection:
            try:
                self.connection.close()
            except Exception:
                pass
